edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,326 @@
+        }
+
+        void InsertDeviceIdentity(ServiceIdentity device)
+        {
+            // Root device is the base-case for constructing the authchain,
+            // child devices derive their authchain from the parent
+            var newNode = device.Id == this.actorDeviceId ?
+                new ServiceIdentityTreeNode(device, Option.Some(this.actorDeviceId)) :
+                new ServiceIdentityTreeNode(device, Option.None<string>());
+
+            this.nodes.Add(device.Id, newNode);
+
+            // Check if there's an existing parent for this device
+            foreach (string parentScopeId in device.ParentScopes)
+            {
+                // Look for the parent based on scope ID, it's okay if we can't find a parent
+                // node. It's possible that the identity for the parent hasn't came in yet.
+                // In this case we'll just leave the new identity as a dangling node, and
+                // the parent device identity should come in at a later point.
+                Option<ServiceIdentityTreeNode> parentNode = this.FindDeviceByScopeId(parentScopeId);
+
+                // Hook up the new node into the tree
+                parentNode.ForEach(p =>
+                {
+                    p.AddChild(newNode);
+                });
+            }
+
+            // Check if there are any dangling child devices that can now be hooked up,
+            // this include placing the new node as the new root.
+            List<ServiceIdentityTreeNode> danglingChildren =
+                this.nodes
+                .Select(kvp => kvp.Value)
+                .Where(s => s.Identity.ParentScopes.Count() > 0 && s.Identity.ParentScopes.Contains(device.DeviceScope.OrDefault()))
+                .ToList();
+
+            // Also check for any modules that should be parented to this new device
+            danglingChildren.AddRange(this.nodes
+                .Select(kvp => kvp.Value)
+                .Where(s => s.Identity.IsModule && s.Identity.DeviceId == device.DeviceId)
+                .ToList()
+                .AsEnumerable());
+
+            foreach (ServiceIdentityTreeNode child in danglingChildren)
+            {
+                newNode.AddChild(child);
+            }
+        }
+
+        void RemoveSingleNode(string id)
+        {
+            if (this.nodes.TryGetValue(id, out ServiceIdentityTreeNode target))
+            {
+                // Unhook from parent
+                target.Parent.ForEach(p => p.RemoveChild(target));
+
+                // Unhook the children
+                target.RemoveAllChildren();
+
+                // Remove the node itself
+                this.nodes.Remove(id);
+            }
+        }
+
+        Option<ServiceIdentityTreeNode> FindDeviceByScopeId(string scopeId)
+        {
+            Preconditions.CheckNonWhiteSpace(scopeId, nameof(scopeId));
+
+            // Look for Edge devices with a matching scope ID
+            List<ServiceIdentityTreeNode> devices =
```

Comment:
the code will prob be more bug prone, generally I'd like to avoid parsing anything if the alternative isn't very expensive.  In this case, I don't think the linear search is that bad since we filter by Edge devices first, and that's just a boolean check.